src/shortestpaths/desopo-pape.jl

```diff
@@ -0,0 +1,49 @@
+"""
+    struct DEposoPapeState{T, U}
+
+An [`AbstractPathState`](@ref) designed for D`Esopo-Pape shortest-path calculations.
+"""
+struct DEsopoPapeState{T <:Real, U <: Integer} <: AbstractPathState
+    parents::Vector{U}
+    dists::Vector{T}
+end
+
+function desopo_pape_shortest_paths(g::AbstractGraph, 
+    src::Integer,
+    distmx::AbstractMatrix{T} = weights(g)) where T <: Real
+    U = eltype(g)
+    nvg = nv(g)
+    dists = fill(typemax(T), nvg)
+    parents = zeros(U, nvg)
+    state = Vector{UInt8}()
+    for i=1:nvg
+        push!(state, 2)
+    end
+    q = Vector{U}()
+    push!(q, src)
+    @inbounds dists[src] = 0
+    
+    while !isempty(q)
+        @inbounds u = q[1]
+        popfirst!(q)
```

Comment:
I don't know exactly how performat `popfirst!` and `pushfirst!` for vectors are. It might be worth to use a queue from the `DataStructures.jl` package. If you do that, you should benchmark, if that makes your code faster or slower.

Reply:
Also `popfirst!` does also return the element that pops, so you can shorten these two lines to one.

Reply:
Both `popfirst!` and `pushfirst!` are `O(1)` operations, so it shouldn't matter if I use a `queue` or not. I'll try benchmarking and update you about the result. 

> Also `popfirst!` does also return the element that pops, so you can shorten these two lines to one.

Ok, nice observation!

Reply:
Having `Deque` in the code is slowing it down a lot! I suggest we keep the implementation as it is. Moreover, it has been mentioned in the documentation of `Deque`, 
>Benchmark shows that the performance of Deque is comparable to Vector on push!, and is noticeably faster on unshift! (by about 30% to 40%).